lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -660,6 +644,108 @@ bool mlir::triton::preProcessLoopAndGetSchedule(
   return true;
 }
 
+/// Find the minimum number of async_commit_group ops between the extract
+/// and the insert. Wait number is the number of commits-1.
+static int minWaitNumberForExtract(ttg::ExtractSliceOp extractOp) {
+  auto countCommitsBetween = [](Operation *op1, Operation *op2) {
+    int count = 0;
+    for (auto op = op1; op != op2; op = op->getNextNode()) {
+      if (isa<ttg::AsyncCommitGroupOp>(op))
+        count++;
+      // Intentionally skip block ops' children. This will give us
+      // convervatively low number of insert ops.
+    }
+    return count;
+  };
+
+  int minCommitNumber = INT_MAX;
+
+  // DFS the def chain of the extract op to find the insert op. On each path
+  // we calculate the number of insert ops. Then we select the minimum number
+  // of insert ops among all the paths.
+  // If the wait is not needed (when insert is a TMA insert), we return -1.
+  std::function<int(Value, Operation *, int)> minOverHistories =
+      [&](Value val, Operation *sinkOp, int thisHistorySum) -> int {
+    if (Operation *defOp = val.getDefiningOp()) {
+      if (isa<ttg::InsertSliceAsyncOp>(defOp)) {
+        thisHistorySum += countCommitsBetween(defOp->getNextNode(), sinkOp);
+        minCommitNumber = std::min(minCommitNumber, thisHistorySum);
+        return minCommitNumber;
+      }
+      if (isa<ttng::InsertSliceTMAOp>(defOp)) {
+        // We don't need to wait for TMA inserts.
+        return -1;
+      }
+      // Failed to track, return 1 conservatively.
+      return 1;
+    }
+    if (auto arg = val.dyn_cast<BlockArgument>()) {
+      Block *block = arg.getOwner();
+      auto forOp = dyn_cast<scf::ForOp>(block->getParentOp());
+
+      // Failed to track, return 1 conservatively.
+      if (!forOp)
+        return 1;
+
+      Operation *firstForInst = &*forOp.getBody()->begin();
+      int insertsBetween = countCommitsBetween(firstForInst, sinkOp);
+      thisHistorySum += insertsBetween;
+      if (thisHistorySum >= minCommitNumber)
+        return minCommitNumber;
+
+      // get the value value assigned to the argument coming from outside the
+      // loop
+      Value incomingVal = forOp.getInitArgs()[arg.getArgNumber() - 1];
+      int min1 = minOverHistories(incomingVal, forOp, thisHistorySum);
+
+      // get the value value assigned to the argument coming from the previous
+      // iteration
+      Operation *yieldOp = block->getTerminator();
+      Value prevVal = yieldOp->getOperand(arg.getArgNumber() - 1);
+      int min2 = minOverHistories(prevVal, yieldOp, thisHistorySum);
+      return std::min(std::min(min1, min2), minCommitNumber);
+    }
+    // Failed to track, return 1 conservatively.
+    return 1;
+  };
+
+  int minCommits = minOverHistories(extractOp.getOperand(0), extractOp, 0);
+  if (minCommits == -1)
+    return -1;
+  if (minCommits == 0)
+    llvm::report_fatal_error("No commits between insert and extract!");
+  return minOverHistories(extractOp.getOperand(0), extractOp, 0) - 1;
```

Comment:
`return minCommits -1;` ?

Reply:
I'm bit confused between the meaning of `-1`, `0`, and `1`. Seems like all these values have special meaning?

Reply:
`-1`: no wait is required
`0`: Seems to me still no wait is required
`1`: Wait until all groups done.
`i (i > 1)`: Wait until we have `i` groups left.

Reply:
yeah could be worth making it an std::optional<int> instead. `0` is a valid error since there should be a commit between insert and wait otherwise the IR is invalid. `1` seems legit too. for `i` we wait untile `i-1` groups are done?

Reply:
Sure, should be `i-1` groups. Updated.

Reply:
> `return minCommits -1;` ?

Oooops, my bad.



> yeah could be worth making it an std::optional instead. `0` is a valid error since there should be a commit between insert and wait otherwise the IR is invalid. `1` seems legit too. for `i` we wait untile `i-1` groups are done?

Yeah, making -1 into returning an empty optional sounds good to me, and will make the code less magic.

Reply:
@Jokeren as Thomas wrote, all non-negative `i` means that there is that many commits found between insert and extract. If we cannot track, we pretend that there is 1 commit to be conservative and correct (unless there would turn out to be no commits in reality, but in this case it is a compiler error and we cannot help it anyway).
-1 is a special value. I'll look into replacing it with `std::nullopt`. Does it make sense to you?

Reply:
Sure, thanks!

Reply:
this discussion also points to a subtle bug that was there in the code: When iterating over a group of the extracts, I was selecting the min of all the consecutive extracts and adding wait with that number. If there was an extract that would return `-1`, all the extracts would be considered as not needing a wait (because `-1` would be selected as a minimum). Instead, this single extract should be disregarded for the group.